server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -203,7 +208,8 @@ public DataStream(
         boolean failureStore,
         List<Index> failureIndices,
         boolean rolloverOnWrite,
-        @Nullable DataStreamAutoShardingEvent autoShardingEvent
+        @Nullable DataStreamAutoShardingEvent autoShardingEvent,
+        long failureStoreGeneration
```

Comment:
Nit: I would prefer to move the `failureStoreGeneration` 2 spots earlier in order to group together the arguments relating to `failure_store`. We cannot do this for serialisation but in the constructor signature it could be possible. Again this is a nit, so I just leave it here. 

Same applies to helper methods etc.

Reply:
I actually had that initially 😅. But indeed, for the serialization we need a different order. So, for consistency's sake, I chose to order all constructors in the same way. Would you suggest having a different order for one/some constructors? Or did you have something else in mind exactly?

Reply:
I was on the fence on mentioning the constructor order, but considering that it was something that Mary is pointing out I feel compelled to agree - Maybe we just reorder the constructor parameters and keep the serialization order as is. Then the weirdness is just contained in the serialization code instead of the parameters being separated every time you make a data stream.

Reply:
I will take it as sign that this is a nicer way forward since we were all thinking about it :p